app/main/notify.py

```diff
@@ -1,48 +1,92 @@
-import logging
 from datetime import datetime
+from typing import BinaryIO
 
+from flask import current_app
+from notifications_python_client import prepare_upload
 from notifications_python_client.errors import HTTPError
 from notifications_python_client.notifications import NotificationsAPIClient
+from werkzeug.datastructures import FileStorage
 
 from config import Config
 
 CONFIRMATION_EMAIL_STRFTIME = "%e %B %Y"  # e.g. 1 October 2023
 
 
-def send_confirmation_email(
+def send_email(
     email_address: str,
-    filename: str,
-    place: str,
-    fund: str,
-    reporting_period: str,
-    date_of_submission: datetime = datetime.now(),
+    template_id: str,
+    file: BinaryIO | FileStorage = None,
+    **kwargs,
 ) -> None:
-    """Send confirmation email to the specified email address via the GovUK Notify service.
+    """Send email to the specified email address via the GovUK Notify service.
 
     :param email_address: recipient email address
-    :param filename: filename of the submission
-    :param place: place submitted for
-    :param fund: fund submitted for
-    :param reporting_period: submission reporting period
-    :param date_of_submission: date of submission (defaults to now)
+    :param template_id: template ID of GovNotify template
+    :param file: optional - file to send by email
+    :param kwargs: any kwargs are sent as email personalisation values
     :return: None
     """
-    personalisation = {
-        "name of fund": fund,
-        "place name": place,
-        "reporting period": reporting_period,
-        "filename": filename,
-        "date_of_submission": date_of_submission.strftime(CONFIRMATION_EMAIL_STRFTIME).strip(),
-    }
+    personalisation = kwargs
+    if file:
+        link_file(file, personalisation)
     try:
         notify_client = NotificationsAPIClient(Config.NOTIFY_API_KEY)
         notify_client.send_email_notification(
             email_address=email_address,
-            template_id=Config.CONFIRMATION_EMAIL_TEMPLATE_ID,
+            template_id=template_id,
             personalisation=personalisation,
         )
-        logging.info("Confirmation email sent via GovUK Notify")
+        current_app.logger.info("Email sent via GovUK Notify")
     except HTTPError as notify_exc:
-        logging.error(
-            f"HTTPError when trying to send confirmation email: params={personalisation} exception={notify_exc}"
+        current_app.logger.error(
+            f"HTTPError when trying to send email: params={personalisation} exception={notify_exc}"
         )
+
+
+def link_file(file: FileStorage, personalisation: dict):
+    """Links a file to an email.
+
+    The file size limit is 2MB - files larger than this will not be sent.
+
+    :param file: file to link
+    :param personalisation: email personalisation key value pairs
+    :return: None
+    """
+    file.seek(0)
```

Comment:
Is it possible to make this `BinaryIO` or some other way that doesn't involve seeking the file. 
Could this return optionally the attachment rather than mutating the personalisation dict, which is a bit more opaque?

Reply:
We still have to seek the file somewhere, regardless if it's a `BinaryIO` because it's already been read when sent to `/ingest`. I've moved the seek to straight after the POST is sent. 

I've also changed this function so it's just defined as a wrapper to the existing `prepare_upload` function that has some exception handling and logging on top.